### DISCUSSION TITLE
OpenGL error running on Ubuntu?

### DISCUSSION THREAD

**Discussion by ChrisShawK2** [unoplatform/uno#9097] [Q&A]
I built the basic sample Uno Platform app with Visual Studio 2022 on Windows.  I built for Windows and Skia.Gtk, and ran both successfully from the IDE.  After publishing each, I ran the win-x64 version from the published folder.  I copied the linux-x64 folder to Ubuntu (22.04), and tried to run it there.  What I got was:

`shaw@ubuntu22:~/Desktop/linux-x64$ ./Uno.Skia.Gtk`
`nullptr GL version string.Marshaling render signal`
`GLIB UNHANDLED EXCEPTIONSystem.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.`
 `---> System.NotSupportedException: OpenGL is not supported in this system`
`   at Uno.UI.Runtime.Skia.OpenGLRenderSurface.TryBuildGRContext()`
`   at Uno.UI.Runtime.Skia.GLRenderSurfaceBase.UnoGLDrawingArea_Render(Object o, RenderArgs args)`
`   --- End of inner exception stack trace ---`
`   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span``1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)`
`   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)`
`   at System.Delegate.DynamicInvokeImpl(Object[] args)`
`   at System.Delegate.DynamicInvoke(Object[] args)`
`   at GLib.Signal.ClosureInvokedCB(Object o, ClosureInvokedArgs args)`
`   at GLib.SignalClosure.MarshalCallback(IntPtr raw_closure, IntPtr return_val, UInt32 n_param_vals, IntPtr param_values, IntPtr invocation_hint, IntPtr marshal_data)`

I verified:
`cshaw@ubuntu22:~/Desktop/linux-x64$ glxinfo | grep "OpenGL version"`
`OpenGL version string: 4.1 (Compatibility Profile) Mesa 22.0.1`

What am I missing?

Thanks,
Chris

**Comment by jeromelaban** [comment]
Thanks for the report! Are you running on an actual Linux machine, or in WSL? Also, which version of Uno are you running with?

**Reply by ChrisShawK2** [reply]
Actually, neither.  I'm running Ubuntu in a VMWare VM (they won't buy me a new machine for testing...).  The Uno I'm running with is what I installed today (after getting the VS 2022 Preview...  ;-) ).  I'm trying to write a Software Licensing package that can be used for both Windows and Linux applications, and uno looked like my best bet.

**Reply by jeromelaban** [reply]
Got it! So there may be something specific with VMWare's implementation of OpenGL that we have not found yet. 

Which version of Uno are you using in your Skia.Gtk csproj?

**Comment by ChrisShawK2** [comment]
![image](https://user-images.githubusercontent.com/39476187/175184632-f5d02512-557a-4811-96e7-2f9341a37da2.png)

**Comment by jeromelaban** [comment]
Thanks. The location of the error suggest that while OpenGL was detected as useable, Skia does not find it useable meaning that our detection algorithms need improvements :)

You can change the render surface temporarily with this: https://platform.uno/docs/articles/features/using-skia-gtk.html?q=skia#changing-the-rendering-target

Still, let's track why the detection fails with https://github.com/unoplatform/uno/issues/9099

**Reply by ChrisShawK2** [reply]
OK, I'll try changing it, and watch that issue.  Let me know if you need any more info from me (though it's basically *your* code, so you'd probably not find much that you didn't recognize...)

For today, I really just wanted to see if I can build something on Windows that I can run on a Linux 'machine'.  I used the VM because I don't trust WSL to be "Windows-free".  (It doesn't help if I make it work there, but it still won't run on a customer's machine.)

**Reply by jeromelaban** [reply]
That's fair. If you have the ability to run on an actual Linux machine, to validate if it's not something else we've missed, it'll help.

Also, are you able to run other OpenGL based software on that VM?

**Reply by ChrisShawK2** [reply]
Don't know -- I just built it today ;-)  Mesa found and recognized it, though.  Plus, I ran your Win Calculator example on it.  (I didn't build that, though).

**Reply by ChrisShawK2** [reply]
...and, I can't lay hands on a real Linux box just now, but I'll ask around...

**Reply by ChrisShawK2** [reply]
FYI:  I tried adding the line (as suggested):
`host.RenderSurfaceType = RenderSurfaceType.Software;`

..and got *EXACTLY* the same error trace.

**Reply by jeromelaban** [reply]
This seems unlikely, are you sure to have rebuilt the app completely?

To be sure, I've validated again using an Ubuntu 22.04 running in Hyper-V, and a default application (using `dotnet new unoapp`) starts properly. We'll likely need to troubleshoot the code paths more precisely. Could you share the configuration parameters of your VM?

**Reply by ChrisShawK2** [reply]
Agreed -- I thought it unlikely as well -- that's why I did it twice before I responded.  And again just now.

I feel that something else is going on, but I'm new here.  What else could I be missing?

I've tried to keep this as simple as possible.  I didn't use  `dotnet new unoapp` -- I followed the steps [here](https://platform.uno/docs/articles/get-started-vs-2022.html?tabs=windows),  I selected *only*  Windows and Gtk, but otherwise, did as you said.

**Reply by jeromelaban** [reply]
Getting some additional information about your VM would help. I'm not familiar with recent builds of VMWare, but it seems that 3D acceleration [needs to be explicitly enabled somehow](https://docs.vmware.com/en/VMware-Workstation-Pro/15.0/com.vmware.ws.using.doc/GUID-FF434E5C-2FEE-48C9-BEAF-943F0536301E.html). Is that the case in your environment?

**Reply by ChrisShawK2** [reply]
I defaulted everything when I created the VM -- if anything else was necessary, I didn't do it.  Here's what I've got:

![image](https://user-images.githubusercontent.com/39476187/175203314-21243c60-d804-403d-b2e6-341bcc45dab6.png)
![image](https://user-images.githubusercontent.com/39476187/175203364-fc20d080-5204-4f82-93ce-ed4dd4a8e53a.png)

Is 3D acceleration really needed for a TextBlock and a Slider?

**Reply by jeromelaban** [reply]
Most UI frameworks use OpenGL (or similar) acceleration by default nowadays :) 

But no, it's not needed as there's a software renderer mode for that, but for some reason the override does not seem to work on your end. But until there's a fix to get Uno working on VMWare, the software renderer will need to be used.

I'd suggest trying the same modification you made on windows directly on the ubuntu machine, just to rule out file copy to the VM issues.

**Reply by ChrisShawK2** [reply]
OK, desperate measures:
After deleting the entire \bin\ directory, and *then* rebuilding (with the RenderSurface changed) and then republishing, the app ran on Ubuntu!  (Note that 'Rebuild All' didn't achieve this, nor did 'Clean' followed by 'Rebuild'.)

So, at least *that's* working -- I'll still wait to see about the OpenGL part.

Thanks!
--Chris

**Reply by jeromelaban** [reply]
I've posted [an update](https://github.com/unoplatform/uno/issues/9099#issuecomment-1167970257) on the OpenGL part, which may be of use on your side. Let us know if this helps.

**Reply by ChrisShawK2** [reply]
After adding the package you mentioned, commenting out the RenderSurfaceType fix and then rebuilding, the app ran fine.
Thanks!

**Reply by jeromelaban** [reply]
Thanks for the update! We've updated the documentation to account for this, but we'll still need to change our rendering pipeline to avoid it completely.

**Comment by LiphiTC** [comment]
Hello, I'm getting kinda the same error on native Arch linux.
With Intel mesa
![image](https://user-images.githubusercontent.com/48130941/181387866-485533d9-b6b1-41e1-8eed-b5e8df725ddb.png)
And with Nvidia proprietary driver
![image](https://user-images.githubusercontent.com/48130941/181388376-6dbe37e3-3ae7-438b-bfd0-9e507555f0eb.png)

**Reply by LiphiTC** [reply]
Turns out that Nvidia driver had some issue with opengl. Now it has the same output as Intel

**Reply by megaheart** [reply]
Add the line at the file `Program.cs` in project `<YourProjectName>.Skia.Gtk`:
```host.RenderSurfaceType = RenderSurfaceType.Software;```

I do as the following and now my uno app can run in WSL
